### PR TITLE
Adjust System.Net.Http metrics

### LINKED ISSUE #88515: [Design] .NET metrics naming standard
What should instruments and tags be named?

Options:
1. Follow event counters naming standard
2. Follow OTel semantic conventions naming standard

Needs to be decided before .NET 8 ships.

### LINKED ISSUE #89093: Adjust Networking metrics to follow OpenTelemetry semantic conventions
We are implementing the following metrics for .NET 8 with ad-hoc naming:

```
// Request metrics:
http-client-current-requests (scheme, host, port, method)
http-client-request-duration (scheme, host, port, method, status-code, protocol)
http-client-failed-requests  (scheme, host, port, method, status-code, protocol)

// Connection metrics:
http-client-current-connections      (protocol, scheme, host, port)
http-client-current-idle-connections (protocol, scheme, host, port)
http-client-connection-duration      (protocol, scheme, host, port)

// How long is a request waiting in the connection pool's queue:
http-client-requests-queue-duration  (protocol, scheme, host, port)

dns-lookups-requested(hostname)
```

We need to adjust these instruments to follow [OpenTelemetry semantic conventions](https://opentelemetry.io/docs/specs/otel/metrics/semantic_conventions/) in naming and other aspects. Details TBD.

### LINKED ISSUE #89451: Test failure: HttpMetricsTest_Http*.CurrentRequests_InstrumentEnabledAfterSending_NotRecorded
### Error Blob

```json
{
  "ErrorMessage": "CurrentRequests_InstrumentEnabledAfterSending_NotRecorded",
  "BuildRetry": false,
  "ErrorPattern": "",
  "ExcludeConsoleLog": false
}
```

### Reproduction Steps

_No response_

 ### Known issue validation
**Build: :mag_right:** https://dev.azure.com/dnceng-public/public/_build/results?buildId=349185
**Error message validated:** `CurrentRequests_InstrumentEnabledAfterSending_NotRecorded`
**Result validation: :white_check_mark:** Known issue matched with the provided build.
**Validation performed at:** 7/25/2023 4:52:14 PM UTC

### Report

|Build|Definition|Test|Pull Request|
|---|---|---|---|
|[361944](https://dev.azure.com/dnceng-public/public/_build/results?buildId=361944)|dotnet/runtime|[System.Net.Http.Functional.Tests.HttpMetricsTest_Http20.CurrentRequests_InstrumentEnabledAfterSending_NotRecorded](https://dev.azure.com/dnceng-public/public/_build/results?buildId=361944&view=ms.vss-test-web.build-test-results-tab&runId=7645334&resultId=104088)||
|[361941](https://dev.azure.com/dnceng-public/public/_build/results?buildId=361941)|dotnet/runtime|[System.Net.Http.Functional.Tests.HttpMetricsTest_Http20.CurrentRequests_InstrumentEnabledAfterSending_NotRecorded](https://dev.azure.com/dnceng-public/public/_build/results?buildId=361941&view=ms.vss-test-web.build-test-results-tab&runId=7645278&resultId=104097)||
|[360399](https://dev.azure.com/dnceng-public/public/_build/results?buildId=360399)|dotnet/runtime|[System.Net.Http.Functional.Tests.SocketsHttpHandler_HttpClientHandler_Http11_Cancellation_Test.ConnectionFailure_AfterInitialRequestCancelled_SecondRequestSucceedsOnNewConnection](https://dev.azure.com/dnceng-public/public/_build/results?buildId=360399&view=ms.vss-test-web.build-test-results-tab&runId=7605914&resultId=104179)||
|[360395](https://dev.azure.com/dnceng-public/public/_build/results?buildId=360395)|dotnet/runtime|[System.Net.Http.Functional.Tests.HttpMetricsTest_Http20.CurrentRequests_InstrumentEnabledAfterSending_NotRecorded](https://dev.azure.com/dnceng-public/public/_build/results?buildId=360395&view=ms.vss-test-web.build-test-results-tab&runId=7605832&resultId=104092)||
|[360393](https://dev.azure.com/dnceng-public/public/_build/results?buildId=360393)|dotnet/runtime|[System.Net.Http.Functional.Tests.HttpMetricsTest_Http20_HttpMessageInvoker.CurrentRequests_InstrumentEnabledAfterSending_NotRecorded](https://dev.azure.com/dnceng-public/public/_build/results?buildId=360393&view=ms.vss-test-web.build-test-results-tab&runId=7605810&resultId=104091)||
|[358694](https://dev.azure.com/dnceng-public/public/_build/results?buildId=358694)|dotnet/runtime|[System.Net.Http.Functional.Tests.HttpMetricsTest_Http11_Sync.CurrentRequests_InstrumentEnabledAfterSending_NotRecorded](https://dev.azure.com/dnceng-public/public/_build/results?buildId=358694&view=ms.vss-test-web.build-test-results-tab&runId=7556180&resultId=104093)||
|[358679](https://dev.azure.com/dnceng-public/public/_build/results?buildId=358679)|dotnet/runtime|[System.Net.Http.Functional.Tests.HttpMetricsTest_Http11_Sync.CurrentRequests_InstrumentEnabledAfterSending_NotRecorded](https://dev.azure.com/dnceng-public/public/_build/results?buildId=358679&view=ms.vss-test-web.build-test-results-tab&runId=7556156&resultId=104090)||
|[358700](https://dev.azure.com/dnceng-public/public/_build/results?buildId=358700)|dotnet/runtime|[System.Net.Http.Functional.Tests.HttpMetricsTest_Http20.CurrentRequests_InstrumentEnabledAfterSending_NotRecorded](https://dev.azure.com/dnceng-public/public/_build/results?buildId=358700&view=ms.vss-test-web.build-test-results-tab&runId=7556178&resultId=104093)||
|[356701](https://dev.azure.com/dnceng-public/public/_build/results?buildId=356701)|dotnet/runtime|[System.Net.Http.Functional.Tests.HttpMetricsTest_Http11_Async.CurrentRequests_InstrumentEnabledAfterSending_NotRecorded](https://dev.azure.com/dnceng-public/public/_build/results?buildId=356701&view=ms.vss-test-web.build-test-results-tab&runId=7506040&resultId=104085)||
|[356699](https://dev.azure.com/dnceng-public/public/_build/results?buildId=356699)|dotnet/runtime|[System.Net.Http.Functional.Tests.SocketsHttpHandler_PostScenarioTest.PostNonRewindableContentUsingAuth_PreAuthenticate_Success](https://dev.azure.com/dnceng-public/public/_build/results?buildId=356699&view=ms.vss-test-web.build-test-results-tab&runId=7505832&resultId=104190)||
|[356668](https://dev.azure.com/dnceng-public/public/_build/results?buildId=356668)|dotnet/runtime|[System.Net.Http.Functional.Tests.HttpMetricsTest_Http11_Async_HttpMessageInvoker.CurrentRequests_InstrumentEnabledAfterSending_NotRecorded](https://dev.azure.com/dnceng-public/public/_build/results?buildId=356668&view=ms.vss-test-web.build-test-results-tab&runId=7505802&resultId=104091)||
|[356198](https://dev.azure.com/dnceng-public/public/_build/results?buildId=356198)|dotnet/runtime|[System.Net.Http.Functional.Tests.HttpMetricsTest_Http11_Async.CurrentRequests_InstrumentEnabledAfterSending_NotRecorded](https://dev.azure.com/dnceng-public/public/_build/results?buildId=356198&view=ms.vss-test-web.build-test-results-tab&runId=7492500&resultId=104092)||
|[355091](https://dev.azure.com/dnceng-public/public/_build/results?buildId=355091)|dotnet/runtime|[System.Net.Http.Functional.Tests.SocketsHttpHandler_HttpClientHandler_Http11_Cancellation_Test.ConnectionFailure_AfterInitialRequestCancelled_SecondRequestSucceedsOnNewConnection](https://dev.azure.com/dnceng-public/public/_build/results?buildId=355091&view=ms.vss-test-web.build-test-results-tab&runId=7466982&resultId=104096)||
|[355081](https://dev.azure.com/dnceng-public/public/_build/results?buildId=355081)|dotnet/runtime|[System.Net.Http.Functional.Tests.HttpMetricsTest_Http20.CurrentRequests_InstrumentEnabledAfterSending_NotRecorded](https://dev.azure.com/dnceng-public/public/_build/results?buildId=355081&view=ms.vss-test-web.build-test-results-tab&runId=7466896&resultId=104092)||
|[353753](https://dev.azure.com/dnceng-public/public/_build/results?buildId=353753)|dotnet/runtime|[System.Net.Http.Functional.Tests.HttpMetricsTest_Http20_HttpMessageInvoker.CurrentRequests_InstrumentEnabledAfterSending_NotRecorded](https://dev.azure.com/dnceng-public/public/_build/results?buildId=353753&view=ms.vss-test-web.build-test-results-tab&runId=7434938&resultId=104089)||
|[353747](https://dev.azure.com/dnceng-public/public/_build/results?buildId=353747)|dotnet/runtime|[System.Net.Http.Functional.Tests.HttpClientHandler_RemoteServerTest.GetAsync_SetAutomaticDecompression_ContentDecompressed_Deflate](https://dev.azure.com/dnceng-public/public/_build/results?buildId=353747&view=ms.vss-test-web.build-test-results-tab&runId=7434942&resultId=104138)||
|[349185](https://dev.azure.com/dnceng-public/public/_build/results?buildId=349185)|dotnet/runtime|[System.Net.Http.Functional.Tests.SocketsHttpHandler_HttpClientHandler_Http11_Cancellation_Test.ConnectionFailure_AfterInitialRequestCancelled_SecondRequestSucceedsOnNewConnection](https://dev.azure.com/dnceng-public/public/_build/results?buildId=349185&view=ms.vss-test-web.build-test-results-tab&runId=7322724&resultId=104104)||
#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|2|14|17|

### PR DESCRIPTION
**Pull request by antonfirsov**
Adjust System.Net.Http metrics naming and semantics according to the outcome of the discussion in https://github.com/lmolkova/semantic-conventions/pull/1:

https://github.com/lmolkova/semantic-conventions/blob/dotnet8-metrics/docs/dotnet/dotnet-http-metrics.md

Contributes to #89093.
Fixes #89451.

### CODE REVIEW

File: src/libraries/System.Net.Http/tests/FunctionalTests/MetricsTest.cs [hunk 1]
```diff
@@ -205,51 +249,67 @@ public HttpMetricsTest(ITestOutputHelper output) : base(output)
...
-                using InstrumentRecorder<long> recorder = new(Handler.MeterFactory, InstrumentNames.CurrentRequests);
-                instrumentEnabledSemaphore.Release();
-
+                Task<HttpResponseMessage> clientTask = Task.Run(() => SendAsync(client, request));
+                await connectionStarted.Task;
+                using InstrumentRecorder<long> recorder = new(Handler.MeterFactory, InstrumentNames.ActiveRequests);
                 using HttpResponseMessage response = await clientTask;
 
                 Assert.Empty(recorder.GetMeasurements());
             }, async server =>
```
Comment (antonfirsov) [inline src/libraries/System.Net.Http/tests/FunctionalTests/MetricsTest.cs:311]: This is a fix for #89451 by making the test deterministic.

**Review by MihaZupan — APPROVED** [review]

File: src/libraries/System.Net.Http/src/System/Net/Http/Metrics/MetricsHandler.cs [hunk 2]
```diff
@@ -22,21 +22,18 @@ public MetricsHandler(HttpMessageHandler innerHandler, IMeterFactory? meterFacto
...
+                "http.client.active_requests",
                 description: "Number of outbound HTTP requests that are currently active on the client.");
-            _failedRequests = meter.CreateCounter<long>(
-                "http-client-failed-requests",
-                description: "Number of outbound HTTP requests that have failed.");
             _requestsDuration = meter.CreateHistogram<double>(
-                "http-client-request-duration",
+                "http.client.request.duration",
                 unit: "s",
                 description: "The duration of outbound HTTP requests.");
```
[the remark this answers is not in the document]
Comment (antonfirsov) [reply src/libraries/System.Net.Http/src/System/Net/Http/Metrics/MetricsHandler.cs:31]: I didn't like the "Measures" verb here 😄, trying to fix the doc instead: https://github.com/lmolkova/semantic-conventions/pull/1#pullrequestreview-1559584159

**Comment by antonfirsov** [comment]
All CI failures are unrelated.

File: src/libraries/System.Net.Http/src/System/Net/Http/Metrics/MetricsHandler.cs [hunk 3]
```diff
@@ -83,67 +80,90 @@ protected override void Dispose(bool disposing)
...
+        private static string GetProtocolVersionString(Version httpVersion) => (httpVersion.Major, httpVersion.Minor) switch
         {
-            (1, 0) => "HTTP/1.0",
-            (1, 1) => "HTTP/1.1",
-            (2, 0) => "HTTP/2",
-            (3, 0) => "HTTP/3",
-            _ => $"HTTP/{httpVersion.Major}.{httpVersion.Minor}"
+            (1, 0) => "1.0",
+            (1, 1) => "1.1",
+            (2, 0) => "2.0",
```
Comment (lmolkova) [inline src/libraries/System.Net.Http/src/System/Net/Http/Metrics/MetricsHandler.cs:164]: I think @JamesNK uses "2" and "3" for HTTP/2 and HTTP/3 for asp.net core and kestrel, which, I believe, matches their RFCs ([HTTP/2](https://datatracker.ietf.org/doc/html/rfc7540), [HTTP/3](https://www.rfc-editor.org/info/rfc9114)).

can we still unify?

Comment (antonfirsov) [reply src/libraries/System.Net.Http/src/System/Net/Http/Metrics/MetricsHandler.cs:164]: Yup I will open a follow-up PR.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Metrics/SocketsHttpHandlerMetrics.cs [hunk 4]
```diff
@@ -8,25 +8,21 @@ namespace System.Net.Http.Metrics
...
     {
-        public readonly UpDownCounter<long> CurrentConnections = meter.CreateUpDownCounter<long>(
-            name: "http-client-current-connections",
-            description: "Number of outbound HTTP connections that are currently active on the client.");
-
-        public readonly UpDownCounter<long> IdleConnections = meter.CreateUpDownCounter<long>(
-            name: "http-client-current-idle-connections",
-            description: "Number of outbound HTTP connections that are currently idle on the client.");
+        public readonly UpDownCounter<long> OpenConnections = meter.CreateUpDownCounter<long>(
+            name: "http.client.open_connections",
```
Comment (noahfalk) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Metrics/SocketsHttpHandlerMetrics.cs:12]: The unit should be set to "{connection}"

@lmolkova - confirming that is correct and not just some formatting thing in the spec that wasn't intended to be used literally?

File: src/libraries/System.Net.Http/src/System/Net/Http/Metrics/MetricsHandler.cs [hunk 5]
```diff
@@ -22,21 +22,18 @@ public MetricsHandler(HttpMessageHandler innerHandler, IMeterFactory? meterFacto
             meter = meterFactory?.Create("System.Net.Http") ?? SharedMeter.Instance;
 
             // Meter has a cache for the instruments it owns
-            _currentRequests = meter.CreateUpDownCounter<long>(
-                "http-client-current-requests",
+            _activeRequests = meter.CreateUpDownCounter<long>(
+                "http.client.active_requests",
```
Comment (noahfalk) [inline src/libraries/System.Net.Http/src/System/Net/Http/Metrics/MetricsHandler.cs:26]: unit should be "{request}"

**Review by noahfalk — COMMENTED** [review]
Sorry I was delayed with some other work, this looked good but I noticed two instruments are missing their units.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Metrics/SocketsHttpHandlerMetrics.cs (hunk 4 above)
Comment (antonfirsov) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Metrics/SocketsHttpHandlerMetrics.cs:12]: Opened https://github.com/dotnet/runtime/pull/90020 PTAL.

**Pull request merged**